alns/select/MABSelector.py

```diff
@@ -0,0 +1,216 @@
+            # More than four is OK because we only use the first four.
+            raise ValueError(f"Expected four scores, found {len(scores)}")
+
+        if mab.arms != self.make_arms(num_destroy, num_repair, op_coupling):
+            raise ValueError(
+                "Arms in MAB passed to MABSelector are incorrect. Make sure"
+                + "to generate arms with ``MABSelector.make_arms``"
+            )
+
+        self._mab = mab
+        self._scores = scores
+        self._primed = False
+        self._op_coupling = op_coupling
+
+        def extract_context(state):
+            if context_extractor is None:
+                return None
+            else:
+                context = context_extractor(state)
+                if isinstance(context, list):
+                    # if the output is a list, wrap it so it's 2D. Otherwise,
+                    # it's an np array or dataframe, which can be left alone.
+                    context = [context]
+                return context
+
+        self._context_extractor = extract_context
+
+    @property
+    def scores(self) -> List[float]:
+        return self._scores
+
+    @property
+    def mab(self) -> MAB:
+        return self._mab
+
+    @staticmethod
+    def _operators_to_arm(destroy_idx: int, repair_idx: int) -> str:
+        """
+        Converts a tuple of destroy and repair operator indices to an arm
+        string that can be passed to self._mab.
+
+        Examples
+        --------
+        >>> MABSelector._operators_to_arm(0, 1)
+        "0_1"
+        >>> MABSelector._operators_to_arm(12, 3)
+        "12_3"
+        """
+        return f"{destroy_idx}_{repair_idx}"
+
+    @staticmethod
+    def _arm_to_operators(arm: str) -> Tuple[int, int]:
+        """
+        Converts an arm string returned from self._mab to a tuple of destroy
+        and repair operator indices.
+
+        Examples
+        --------
+        >>> MABSelector._arm_to_operators("0_1")
+        (0, 1)
+        >>> MABSelector._arm_to_operators("12_3")
+        (12, 3)
+        """
+        [destroy, repair] = arm.split("_")
+        return int(destroy), int(repair)
+
+    def __call__(
+        self, rnd_state: RandomState, best: State, curr: State
+    ) -> Tuple[int, int]:
+        """
+        Returns the (destroy, repair) operator pair from the underlying MAB
+        strategy
+        """
+        if not self._primed:
+            # Default: return the first allowed operator pair
+            if self._op_coupling is None:
+                return 0, 0
+            else:
+                first_non_zero = np.argmax(self._op_coupling)
+                as_indices = np.unravel_index(
+                    first_non_zero, self._op_coupling.shape
+                )
+                return as_indices
+        else:
+            prediction = self._mab.predict(
+                contexts=self._context_extractor(curr)
+            )
+            return self._arm_to_operators(prediction)
```

Comment:
If my comment on `update()` works, I think we can get rid of the `_primed` flag. Here, a case could be made to just try MABWiser, with a good fall-back in case we run into issues. In particular:
```python
try:
  prediction = self._mab.predict(
    contexts=self._context_extractor(curr)
  )
  return self._arm_to_operators(prediction)
except Exception:
  # This can happen when the MAB object has not yet been fit on any 
  # observations. In that case we return any feasible operator index 
  # pair as a first observation.
  allowed = np.argwhere(self._op_coupling)
  idx = rnd_state.randint(len(allowed))
  return tuple(allowed[idx])
```
Using exceptions for flow control isn't great but I think in this case it is a lot more elegant than maintaining a flag with branches that we need to handle in two separate methods.